packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -121,33 +121,58 @@ export const withBlockBindingSupport = createHigherOrderComponent(
 
 			const attributes = {};
 
-			for ( const [ attributeName, boundAttribute ] of Object.entries(
+			const bindingsBySource = new Map();
+
+			for ( const [ attributeName, binding ] of Object.entries(
 				bindings
 			) ) {
-				const source = sources[ boundAttribute.source ];
+				const { source: sourceName, args: sourceArgs } = binding;
+				const source = sources[ sourceName ];
 				if (
-					! source?.getValue ||
+					! source?.getValues ||
 					! canBindAttribute( name, attributeName )
 				) {
 					continue;
 				}
 
-				const args = {
-					registry,
-					context,
-					clientId,
-					attributeName,
-					args: boundAttribute.args,
-				};
-
-				attributes[ attributeName ] = source.getValue( args );
+				bindingsBySource.set( source, {
+					...bindingsBySource.get( source ),
+					[ attributeName ]: {
+						args: sourceArgs,
+					},
+				} );
+			}
 
-				if ( attributes[ attributeName ] === undefined ) {
-					if ( attributeName === 'url' ) {
-						attributes[ attributeName ] = null;
-					} else {
-						attributes[ attributeName ] =
-							source.getPlaceholder?.( args );
+			if ( bindingsBySource.size ) {
+				for ( const [ source, sourceBindings ] of bindingsBySource ) {
+					// Get values in batch if the source supports it.
+					const values = source.getValues( {
+						registry,
+						context,
+						clientId,
+						sourceBindings,
+					} );
```

Comment:
I find the name `sourceBindings` here unclear, especially because we'd be expecting folks to destructure and use it when registering their sources. Maybe we could do `bindingsObject` or `bindingsCollection` instead, something along those lines? I know naming things is hard.

The same goes for other places in this PR where `sourceBindings` is used.

Reply:
What about using just `bindings`?

Reply:
Just `bindings` would be great! But it's already being used in the upper scope on line 104.

I think it's probably worth renaming the value in the upper scope to something like `bindingsData`, that way we can use `bindings` in the public-facing API.

Reply:
I've changed it in [this commit](https://github.com/WordPress/gutenberg/pull/63185/commits/e244e702ef11e6a1b2b1c40b813a5f71ddd97414). I finally used `blockBindings` for the top variable because they are actually the bindings for the block that is being processed.

Let me know what you think. It's fairly simple to change it to a preferred name.

Reply:
Looks good to me 🙌